### PR TITLE
 check for `typemin(T)` in denominator of rational

### LINKED ISSUE #32569: Rational(1, typemin(Int)) comparisons with 0 and 1 deny transitiveness of  ` <`
Maybe that should be solved. It is related to #15225, probably.
Transitivity of `<` should be maintained also by `Rational{Int}`.

```
julia> 1 // typemin(Int)
-1//-9223372036854775808
julia> 1 // typemin(Int) < 0
true
julia> 1 // typemin(Int) < 1
false
```

I propose to throw `ArgumentError` when ` typemin(T)`  is going to become denominator for all bits types ` T`.

### PR DESCRIPTION
**Pull request by simeonschaub**
fix #32569. This is my first PR here, I hope, I didn't miss anything.

### CODE REVIEW

File: base/rational.jl [hunk 1]
```diff
@@ -11,12 +11,16 @@ struct Rational{T<:Integer} <: Real
     den::T
 
     function Rational{T}(num::Integer, den::Integer) where T<:Integer
-        num == den == zero(T) && __throw_rational_argerror(T)
+        num == den == zero(T) && __throw_rational_argerror_zero(T)
+        if T <: Signed && hasmethod(typemin, Tuple{Type{T}})
```
[the remark this answers is not in the document]
Comment (simeonschaub) [reply base/rational.jl]: Yes, that seems reasonable

[the remark this answers is not in the document]
Comment (rfourquet) [reply base/rational.jl]: There is a `Base.hastypemax` method, falling back on `hasmehtod`, I suggest that you create  similarly a `hastypemin` method here. Being a bits type is unrelated.

**Comment by KlausC** [comment]
I am not happy with your use of `isbitstype(T)`; please replace with `Base.hastypemax(T)`. That is also the tenor of @rfourquet's comment, I think.
The former makes use of the reflection interface and produces runtime code like:
` ` `
julia> @code_llvm isbitstype(Int)
;  @ reflection.jl:452 within `isbitstype'
define i8 @julia_isbitstype_16522(%jl_value_t addrspace(10)*) {
top:
; ┌ @ Base.jl:15 within `getproperty'
   %1 = load i8, i8* inttoptr (i64 140103251893833 to i8*), align 1
   %2 = and i8 %1, 1
; └
  ret i8 %2
}
` ` `
while the proposed call evaluates essentially to a constant at run time:
` ` `
julia> @code_llvm Base.hastypemax(Int)
;  @ intfuncs.jl:726 within `hastypemax'
define i8 @julia_hastypemax_16523(%jl_value_t addrspace(10)*) {
top:
  ret i8 1
}
` ` `

**Comment by simeonschaub** [comment]
Are you suggesting that I implement a `hastypemin` function, analogous to `Base.hastypemax`, or should I just assume that every Integer that  has `typemax` defined also has `typemin` defined and just use `hastypemax` here?

**Comment by rfourquet** [comment]
> Are you suggesting that I implement a hastypemin function, analogous to Base.hastypemax

Yes

**Comment by KlausC** [comment]
The implementation: `hastypemin(x::Type{<:Integer}) = hastypemax(x)` would cover all subtypes of `Integer` defined in `Base` correctly and with the same performance advantage.

BTW: `Base.hastypemax` is not implemented optimally for floating point types.

**Comment by rfourquet** [comment]
> The implementation: hastypemin(x::Type{<:Integer}) = hastypemax(x) would cover all subtypes of Integer defined in Base correctly

But it's not valid in general, the definitions are tiny, let's not try to save duplication here at the cost of an approximation

**Comment by simeonschaub** [comment]
Bump.

**Comment by KlausC** [comment]
bump!
The rationals label is missing.

File: base/rational.jl [hunk 2]
```diff
@@ -11,12 +11,16 @@ struct Rational{T<:Integer} <: Real
     den::T
 
     function Rational{T}(num::Integer, den::Integer) where T<:Integer
-        num == den == zero(T) && __throw_rational_argerror(T)
+        num == den == zero(T) && __throw_rational_argerror_zero(T)
+        if T <: Signed && hastypemin(T)
+            den == typemin(T) &&  __throw_rational_argerror_typemin(T)
```
Comment (rfourquet) [inline base/rational.jl]: I'm sorry, in my previous review I was just looking at the `typemin` stuff, but actually I think this error logic here is not appropriate in general. A type could have a `typemin` without the current definition being an error. The real problem occurs when `typemin(T) == -typemin(T)` (which is typically true for bit types which use 2-complement arithmetic). So I suggest instead moving the error checking in the below line, something like: `T <: Signed && signbit(den) && signbit(-den) && throw(...)`.

Comment (simeonschaub) [reply base/rational.jl]: Yes, makes sense. So just delete all the `hastypemin` stuff?

Comment (KlausC) [reply base/rational.jl]: I don't think, this check is very good, because it requires construction of `-den`, which is expensive in the case of `BigInt`.

**Comment by KlausC** [comment]
I think the previous commit was better, because it does not require to construct the negative of a `BigInt` in that case. Timing examples:

```
julia> @btime Base.hastypemax(BigInt)
  0.016 ns (0 allocations: 0 bytes)
false

julia> x = big(51)^1000;
julia> @btime -$x;
  119.650 ns (3 allocations: 744 bytes)
```

File: base/rational.jl (hunk 2 above)
Comment (rfourquet) [reply base/rational.jl]: It can be constructed only when necessary.

File: base/rational.jl [hunk 3]
```diff
@@ -11,12 +11,17 @@ struct Rational{T<:Integer} <: Real
     den::T
 
     function Rational{T}(num::Integer, den::Integer) where T<:Integer
-        num == den == zero(T) && __throw_rational_argerror(T)
+        num == den == zero(T) && __throw_rational_argerror_zero(T)
+        # issue #32569
+        if T <: Signed
+            signbit(den) && signbit(-den) && __throw_rational_argerror_typemin(T)
+        end
         num2, den2 = signbit(den) ? divgcd(-num, -den) : divgcd(num, den)
```
Comment (rfourquet) [inline base/rational.jl]: As @KlausC suggested, this constructs unnecessarily `-den` once more than necessary. You can rearrange the code to fix that, e.g.
```julia
num2, den2 = if signbit(den)
    den = -den
    T <: Signed && signbit(den) && __throw_rational_argerror_typemin(T)
    divgcd(-num, den)
else
    divgcd(num, den)
end

**Comment by KlausC** [comment]
a little polished:
```
if T<:Signed && signbit(den)
    den = -den
    signbit(den) && __throw_rational_argerror_typemin(T)
    num = -num
end
divgcd(num, den)
```

**Comment by KlausC** [comment]
@StefanKarpinski , what do you think?

File: base/rational.jl [hunk 4]
```diff
@@ -11,12 +11,19 @@ struct Rational{T<:Integer} <: Real
     den::T
 
     function Rational{T}(num::Integer, den::Integer) where T<:Integer
-        num == den == zero(T) && __throw_rational_argerror(T)
-        num2, den2 = signbit(den) ? divgcd(-num, -den) : divgcd(num, den)
-        new(num2, den2)
+        num == den == zero(T) && __throw_rational_argerror_zero(T)
+        # issue #32569
```
Comment (rfourquet) [inline base/rational.jl]: Usually we don't insert references to issues in code (as opposed to tests), and here I think we can understand the reason of this test by following the thrown error, faster than by looking up github. No really a problem though, and probably not necessary to update your PR.

**Review by rfourquet — APPROVED** [review]

File: base/rational.jl (hunk 4 above)
Comment (simeonschaub) [reply base/rational.jl]: That sounds reasonable. Went ahead and deleted it now.

**Comment by KlausC** [comment]
Additional issue? Are these test cases ok now?
```
@test -2 // typemin(Int) == 1//4611686018427387904
@test 2 // typemin(Int) == -1//4611686018427387904
```
maybe the `divgcd` has to be done before the checks.

**Comment by KlausC** [comment]
@simeonschaub, would you mind to insert the changes to fix - last comment?

**Comment by KlausC** [comment]
I am not sure, if the `@show ..` should be there.

The test cases are only valid if `Int == Int64`; should be formulated generically.

**Comment by simeonschaub** [comment]
Oh, sorry. My bad! I pushed a bit too prematurely.

**Comment by rfourquet** [comment]
Looks like good to go, I will merge in one day or two if no other review comments (does it look good for you now @KlausC ?)

**Comment by rfourquet** [comment]
Thanks @simeonschaub for the fix and @KlausC for the review!

**Pull request merged**